### PR TITLE
Actor checkpointing with object lineage reconstruction

### PR DESCRIPTION
**Pull request by stephanie-wang**
Actor checkpointing built on top of the existing reconstruction infrastructure.

Briefly, at every task interval for a given actor, a mock task responsible for saving and resuming checkpoints is submitted. The task tries to save a checkpoint to Redis during normal execution. During reconstruction, the task tries to resume from the most recent checkpoint. If the most recent checkpoint is older, then the task requests reconstruction of the preceding tasks.

### CODE REVIEW

File: src/common/task.cc [hunk 1]
```diff
@@ -217,7 +217,18 @@ ActorID TaskSpec_actor_id(TaskSpec *spec) {
 int64_t TaskSpec_actor_counter(TaskSpec *spec) {
   CHECK(spec);
   auto message = flatbuffers::GetRoot<TaskInfo>(spec);
-  return message->actor_counter();
+  int64_t actor_counter = message->actor_counter();
+  if (actor_counter < 0) {
```
Comment (pcmoritz) [inline src/common/task.cc]: why not "return std::abs(message->actor_counter());"?

File: src/local_scheduler/local_scheduler_algorithm.cc [hunk 2]
```diff
@@ -42,6 +42,7 @@ typedef struct {
    *  restrict the submission of tasks on actors to the process that created the
    *  actor. */
   int64_t task_counter;
+  int64_t assigned_task_counter;
```
Comment (pcmoritz) [inline src/local_scheduler/local_scheduler_algorithm.cc]: can you document this?

**Review by pcmoritz — COMMENTED** [review]
LGTM, can you fix the small comments and rebase it?

**Comment by stephanie-wang** [comment]
Great, thanks! I pushed the fixes.

**Comment by robertnishihara** [comment]
#605

File: src/common/task.cc [hunk 3]
```diff
@@ -217,7 +217,14 @@ ActorID TaskSpec_actor_id(TaskSpec *spec) {
 int64_t TaskSpec_actor_counter(TaskSpec *spec) {
   CHECK(spec);
   auto message = flatbuffers::GetRoot<TaskInfo>(spec);
-  return message->actor_counter();
+  return std::abs(message->actor_counter());
```
Comment (robertnishihara) [inline src/common/task.cc]: Why `std::abs`? Is this negative sometimes? And if so, what does that mean?

File: src/common/task.cc [hunk 4]
```diff
@@ -217,7 +217,14 @@ ActorID TaskSpec_actor_id(TaskSpec *spec) {
...
+}
+
+int64_t TaskSpec_actor_is_checkpoint_method(TaskSpec *spec) {
```
Comment (robertnishihara) [inline src/common/task.cc]: Should this return a `bool`?

File: python/ray/actor.py [hunk 5]
```diff
@@ -40,31 +42,130 @@ def get_actor_method_function_id(attr):
...
+    """
+
+    def actor_method_executor(dummy_return_id, task_counter, actor,
+                              *args):
+        # An actor task's dependency on the previous task is represented by
+        # a dummy argument. Remove this argument before invocation.
+        args = args[:-1]
+        if method_name == "__ray_checkpoint__":
+            # Execute the checkpoint task. NOTE(swang): Checkpoint methods
+            # should not throw an exception.
```
Comment (robertnishihara) [inline python/ray/actor.py]: They definitely could throw an exception, right? E.g., if they try to pickle stuff, pickling can fail. Also, if we allow user-defined checkpointing methods (which may be necessary for neural nets and things like that), the user could have a bug in the checkpointing method.

Does that have any implications for this code? Maybe we should add a test case for the case where checkpointing fails?

File: src/local_scheduler/format/local_scheduler.fbs [hunk 6]
```diff
@@ -36,6 +36,11 @@ enum MessageType:int {
   PutObject
 }
 
+// This message is sent from a worker to a local scheduler.
+table GetTaskRequest {
+  task_success: bool;
```
Comment (robertnishihara) [inline src/local_scheduler/format/local_scheduler.fbs]: What does it mean if `task_success` is True versus False? And why is this necessary?

File: python/ray/actor.py (hunk 5 above)
Comment (stephanie-wang) [reply python/ray/actor.py]: Hmm what do you think the expected behavior should be if the checkpoint task does throw an exception? For both saving and resuming a checkpoint?

For me, I would probably expect saving a checkpoint to store the exception like a normal task, and then continue executing tasks on the actor. An exception while resuming a checkpoint could reconstruct the previous object, as if there were no successful checkpoints. How does that sound?

Yeah, once we figure out the expected behavior, a test case would be good. Right now a checkpoint task that throws an exception would just hang, since it would never put the dummy object.

File: src/local_scheduler/format/local_scheduler.fbs (hunk 6 above)
Comment (stephanie-wang) [reply src/local_scheduler/format/local_scheduler.fbs]: This is a little messy, but it's used to notify the local scheduler of the actor's new task counter. For checkpoint tasks, the actor's task counter should only be updated if the task was successful (the checkpoint was actually saved or resumed). If the checkpoint isn't there, for example, then `task_success` is set to `False` and the local scheduler won't modify the actor's task counter.

File: src/common/task.cc (hunk 3 above)
Comment (stephanie-wang) [reply src/common/task.cc]: I set the counter to be negative for checkpoint tasks. It was mostly to avoid adding another field to the task spec. Let me know if you prefer to just add a field.

File: src/common/task.cc (hunk 4 above)
Comment (stephanie-wang) [reply src/common/task.cc]: Yup, thanks!

File: python/ray/actor.py [hunk 7]
```diff
@@ -214,6 +322,10 @@ def export_actor(actor_id, class_id, actor_method_names, num_cpus, num_gpus,
 
 
 def make_actor(cls, num_cpus, num_gpus, checkpoint_interval):
+    # Add one to the checkpoint interval since we will insert a mock task for
+    # every checkpoint.
+    checkpoint_interval += 1
```
Comment (robertnishihara) [inline python/ray/actor.py]: Not a big deal, but 

```python
@ray.remote(checkpoint_interval=0)
class Foo(object):
    def __init__(self):
        pass
```

causes an infinite loop :)

Comment (stephanie-wang) [reply python/ray/actor.py]: Ah yeah, this is a little unfortunate...

File: python/ray/actor.py [hunk 8]
```diff
@@ -254,9 +372,105 @@ def __ray_restore_from_checkpoint__(cls, pickled_checkpoint):
...
+                    actor_checkpoint_failed = True
+
+            # Fall back to lineage reconstruction if we were unable to load the
+            # checkpoint.
+            if actor_checkpoint_failed:
+                worker.local_scheduler_client.reconstruct_object(
+                    plasma_id.binary())
+                worker.local_scheduler_client.notify_unblocked()
+
+            return actor_checkpoint_failed, error
```
Comment (robertnishihara) [inline python/ray/actor.py]: If I run the following,

```python
import ray
ray.init()

@ray.remote(checkpoint_interval=1)
class Foo(object):
    def __init__(self):
        pass
    def method(self):
        pass
    def __ray_save__(self):
        raise Exception("failure")

f = Foo.remote()
```

Then the first call to 

```python
ray.get(f.method.remote())
```

fails with

```
Traceback (most recent call last):
  File "/Users/rkn/Workspace/ray/python/ray/worker.py", line 728, in _process_task
    *arguments)
  File "/Users/rkn/Workspace/ray/python/ray/actor.py", line 151, in actor_method_executor
    actor_checkpoint_failed, error = method(actor, *args)
  File "/Users/rkn/Workspace/ray/python/ray/actor.py", line 472, in __ray_checkpoint__
    return actor_checkpoint_failed, error
UnboundLocalError: local variable 'error' referenced before assignment
```

And the second call to 

```python
ray.get(f.method.remote())
```

hangs.

Not sure if we want to address that here.

File: test/actor_test.py [hunk 9]
```diff
@@ -1352,6 +1349,93 @@ def __ray_restore__(self, checkpoint):
...
+        # Each inc method has been reexecuted once on the new actor.
+        self.assertEqual(ray.get(actor.get_num_inc_calls.remote()), 101)
+        # Get all of the results that were previously lost. Because the
+        # checkpoints were lost, all methods should be reconstructed.
+        results = ray.get(ids)
+        self.assertEqual(results, list(range(1, 1 + len(results))))
+
+        ray.worker.cleanup()
+
+    def testCheckpointException(self):
```
Comment (robertnishihara) [inline test/actor_test.py]: This test hangs for me, and I see the error

```python
Remote function __ray_checkpoint__ failed with:

Traceback (most recent call last):
  File "/Users/rkn/Workspace/ray/python/ray/worker.py", line 728, in _process_task
    *arguments)
  File "/Users/rkn/Workspace/ray/python/ray/actor.py", line 151, in actor_method_executor
    actor_checkpoint_failed, error = method(actor, *args)
  File "/Users/rkn/Workspace/ray/python/ray/actor.py", line 472, in __ray_checkpoint__
    return actor_checkpoint_failed, error
UnboundLocalError: local variable 'error' referenced before assignment
```

File: python/ray/actor.py [hunk 10]
```diff
@@ -254,9 +372,105 @@ def __ray_restore_from_checkpoint__(cls, pickled_checkpoint):
...
+
+            Returns:
+                A bool representing whether the checkpoint was successfully
+                    loaded (whether the actor can safely execute the next task)
+                    and an Exception instance, if one was thrown.
+            """
+            previous_object_id = previous_object_id[0]
+            # Make sure that a previous object was given.
+            if previous_object_id is None:
+                return False
```
Comment (robertnishihara) [inline python/ray/actor.py]: If this code path is ever taken, that's a bug, right? Or can it happen normally?

File: test/actor_test.py (hunk 9 above)
Comment (robertnishihara) [reply test/actor_test.py]: This is probably a python 3 thing, I think I can fix it.

Comment (robertnishihara) [reply test/actor_test.py]: Ok, fixed it.

**Review by robertnishihara — APPROVED** [review]
Looks good to me!

File: python/ray/actor.py (hunk 10 above)
Comment (stephanie-wang) [reply python/ray/actor.py]: Oh yeah, that's right.

File: src/common/task.cc (hunk 3 above)
Comment (robertnishihara) [reply src/common/task.cc]: instead of negative numbers, i actually think adding a bool to the TaskSpec saying whether it is a checkpoint task or not is cleaner, what do you think?
